### PR TITLE
Make await a template

### PR DESCRIPTION
**Pull request by alehander92**
Adapted  https://github.com/status-im/nim-chronos/pull/47 for asyncdispatch with support for multisync : credit for the original impl to @yglukhov !

### CODE REVIEW

**Comment by alehander42** [comment]
it also helps https://github.com/nim-lang/Nim/issues/11911#issuecomment-519410464 and adding stuff like waitFor check

**Comment by Araq** [comment]
What does that do? Enable ``await`` within templates?

**Comment by alehander42** [comment]
Yes, now i can use it in templates : but it also makes await a template which is a cleaner implementation compared to checking for idents (and one can easily implement other stuff like `waitFor` is defined out of {.async.}, but not inside)

File: lib/pure/asyncmacro.nim [hunk 1]
```diff
@@ -440,7 +399,26 @@ macro multisync*(prc: untyped): untyped =
...
+# based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+template await*[T](f: Future[T]): auto =
+  when declared(internalRetFuture):
+    when not declared(internalTmpFuture):
+      var internalTmpFuture: FutureBase
+    internalTmpFuture = f
+    # internalRetFuture.child = internalTmpFuture
+    yield internalTmpFuture
+    if not isNil(internalTmpFuture.error):
+      echo "error"
```
Comment (juancarlospaco) [inline lib/pure/asyncmacro.nim]: Remove the `echo` :grey_question:

File: lib/pure/asyncmacro.nim [hunk 2]
```diff
@@ -440,7 +399,26 @@ macro multisync*(prc: untyped): untyped =
...
   result.add(sync)
 
-proc await*[T](x: T) =
-  ## The 'await' keyword is also defined here for technical
-  ## reasons. (Generic symbol lookup prepass.)
-  {.error: "Await only available within .async".}
+# seems simpler to define it here, than to define it in each multisync
+template await*(f: untyped): untyped =
+  when declared(internalRetFuture):
+    error "Await expects future"
```
Comment (juancarlospaco) [inline lib/pure/asyncmacro.nim]:
```suggestion
    error "Await expects a Future"
```
Capital because `Future` is a Type.

File: lib/pure/asyncmacro.nim [hunk 3]
```diff
@@ -314,7 +314,7 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
                          createFutureVarCompletions(futureVarIdents, nil))
     outerProcBody.add procCb
 
-    # -> return retFuture
+    # -> return internalRetFuture
```
Comment (juancarlospaco) [inline lib/pure/asyncmacro.nim]: This is not the way Nim does comments `#-> ` it should be just `# `. Looks kinda JavaScripty.

Comment (Clyybber) [reply lib/pure/asyncmacro.nim]: The arrow is indicating what the macro will generate. (Also theres a space here between `#` and `->`)  Would you leave our the arrow here, it would look like commented-out code, which it is not.

File: lib/pure/asyncmacro.nim (hunk 2 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: yes, true!

File: lib/pure/asyncmacro.nim (hunk 1 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: sounds reasonable, i don't remember why i had it

File: lib/pure/asyncmacro.nim [hunk 4]
```diff
@@ -440,7 +399,24 @@ macro multisync*(prc: untyped): untyped =
...
+template await*(f: untyped): untyped =
+  when declared(internalRetFuture):
+    error "Await expects a Future"
+  elif not declared(inMultisync):
+    error "Await only available within {.async.}"
+  else:
+    f
+ 
+# based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+template await*[T](f: Future[T]): auto =
```
Comment (mratsim) [inline lib/pure/asyncmacro.nim]: AFAIK we shouldn't use the `auto` return type in the stdlib

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: hm, is there a more complete description why (no problem to fix it)

File: lib/pure/asyncmacro.nim [hunk 5]
```diff
@@ -440,7 +399,24 @@ macro multisync*(prc: untyped): untyped =
...
   result.add(sync)
 
-proc await*[T](x: T) =
-  ## The 'await' keyword is also defined here for technical
-  ## reasons. (Generic symbol lookup prepass.)
-  {.error: "Await only available within .async".}
+# seems simpler to define it here, than to define it in each multisync
+template await*(f: untyped): untyped =
+  when declared(internalRetFuture):
+    error "Await expects a Future"
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: Please expand this:

```nim
error "Await expects a Future[T], got " & $typeof(f)
```

or similar.

File: lib/pure/asyncmacro.nim [hunk 6]
```diff
@@ -440,7 +399,24 @@ macro multisync*(prc: untyped): untyped =
   result.add(asyncSingleProc(asyncPrc))
   result.add(sync)
 
-proc await*[T](x: T) =
-  ## The 'await' keyword is also defined here for technical
-  ## reasons. (Generic symbol lookup prepass.)
-  {.error: "Await only available within .async".}
+# seems simpler to define it here, than to define it in each multisync
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: Not sure what this means, the template doesn't just apply to multisync, does it?

File: lib/pure/asyncmacro.nim (hunk 4 above)
Comment (dom96) [reply lib/pure/asyncmacro.nim]: because `auto` sucks for documentation and means that the template/proc can effectively return anything in the future.

In this case the return type should be trivial to specify completely, just use ``T``, no?

File: lib/pure/asyncmacro.nim [hunk 7]
```diff
@@ -440,7 +399,24 @@ macro multisync*(prc: untyped): untyped =
...
+# based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+template await*[T](f: Future[T]): auto =
+  when declared(internalRetFuture):
+    when not declared(internalTmpFuture):
+      var internalTmpFuture: FutureBase
+    internalTmpFuture = f
+    yield internalTmpFuture
+    if not isNil(internalTmpFuture.error):
+      raise internalTmpFuture.error
+    cast[type(f)](internalTmpFuture).internalRead()
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: The above 3 lines change semantics. Just call `read` instead, also use an explicit `return` if possible.

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: ok, i can change to read

File: lib/pure/asyncmacro.nim (hunk 4 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: it doesnt work for some reason with `T` : expr has no type errors, it really doesnt seem obvious to me why

**Comment by dom96** [comment]
Any updates on this?

**Comment by alehander92** [comment]
sorry, i left some comments but not sure if easily visible: it doesnt work for some reason with T : something like "expr has no type" errors, it really doesnt seem obvious to me why 

this is for the review task to use `T` instead of `auto`

**Comment by dom96** [comment]
That's fine, but there are other comments that need resolving.

**Comment by alehander92** [comment]
ok, i'll try to push a fix for those

File: lib/pure/asyncmacro.nim (hunk 7 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: changed to read

File: lib/pure/asyncmacro.nim (hunk 6 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: i think this might be outdated

File: lib/pure/asyncmacro.nim (hunk 5 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: Ok, but i change all those errors to `await` because thats the name of it, `Await` might be incorrect

File: lib/pure/asyncmacro.nim (hunk 4 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: ok, as we discussed IIRC it can stay with `auto`

**Comment by alehander92** [comment]
ok, the only remaining issue seems to be that `tasyncawait_cyclebreaker.nim` requires me to increase the limit from 1000 to ~150_000 which .. seems strange, have to debug how i break it

**Comment by alehander92** [comment]
EDIT: ok code fixed (dont merge yet: wanted to remove some code) 

otherwise Araq said https://irclogs.nim-lang.org/16-04-2020.html#13:51:11 so i'll leave the test as that (even though i think it might be leaking as it constantly deallocates less)

**Comment by alehander92** [comment]
ok, now it should be ready, also used `error` instead of `assert` after i saw a subsequent @yglukhov  PR to chronos https://github.com/status-im/nim-chronos/pull/48

File: lib/pure/asyncmacro.nim [hunk 8]
```diff
@@ -447,8 +325,28 @@ macro multisync*(prc: untyped): untyped =
...
+# overload for type which is not Future[T], based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+template await*(f: untyped): untyped =
+  when declared(retFuture):
+    static:
+      error "await expects Future[T], got " & $typeof(f)
+  elif not declared(inMultisync):
+    static:
+      error "await only available within {.async.}"
+  else:
+    f
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: From what I can tell this is only here to allow multisync to work. It took me a while to figure this out, can you add a comment to explain this?

File: lib/pure/asyncmacro.nim [hunk 9]
```diff
@@ -447,8 +325,28 @@ macro multisync*(prc: untyped): untyped =
...
   result.add(sync)
-
-proc await*[T](x: T) =
-  ## The 'await' keyword is also defined here for technical
-  ## reasons. (Generic symbol lookup prepass.)
-  {.error: "Await only available within .async".}
+  # echo result.repr
+
+# overload for type which is not Future[T], based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+template await*(f: untyped): untyped =
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: AFAIK `untyped` here will make

```
await:
  if true: echo 24
  echo 123
```

Possible. Can we change to `typed`?

File: lib/pure/asyncmacro.nim [hunk 10]
```diff
@@ -447,8 +325,28 @@ macro multisync*(prc: untyped): untyped =
...
+      error "await only available within {.async.}"
+  else:
+    f
+
+# based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+template await*[T](f: Future[T]): auto =
+  when declared(retFuture):
+    when not declared(internalTmpFuture):
+      var internalTmpFuture: FutureBase
+    internalTmpFuture = f
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: Why do you need this temporary variable? Can't you just yield `f` and call `read` on it too?

File: lib/pure/asyncmacro.nim (hunk 8 above)
Comment (dom96) [reply lib/pure/asyncmacro.nim]: Actually, now that I look at the code above, I would just remove this and keep the `stripAwait` logic in `splitProc`. Adding this `inMultisync` variable into the generated code is more hacky. That way you can also reuse the compiler's type mismatch error messages.

File: lib/pure/asyncmacro.nim [hunk 11]
```diff
@@ -257,7 +178,7 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
     outerProcBody.add(prc.body[0])
 
   # -> var retFuture = newFuture[T]()
-  var retFutureSym = genSym(nskVar, "retFuture")
+  var retFutureSym = ident"retFuture"
```
Comment (dom96) [inline lib/pure/asyncmacro.nim]: Why was this changed? Without `genSym` here if someone declares a `retFuture` of their own then there will be a clash...

File: tests/async/tasyncawait_cyclebreaker.nim [hunk 12]
```diff
@@ -65,7 +65,7 @@ while true:
   GC_collectZct()
 
   let (allocs, deallocs) = getMemCounters()
-  doAssert allocs < deallocs + 1000
+  doAssert allocs < deallocs + 150_000 # changed after await_template, not sure if it isn't too much
```
Comment (dom96) [inline tests/async/tasyncawait_cyclebreaker.nim]: Is this just a way to fix flakiness, or do your changes break this test without this change being introduced?

File: lib/pure/asyncmacro.nim (hunk 9 above)
Comment (Araq) [inline lib/pure/asyncmacro.nim]: This merely happens to work, I think you need ``f: typed`` here.

File: tests/async/tasyncawait_cyclebreaker.nim (hunk 12 above)
Comment (Araq) [inline tests/async/tasyncawait_cyclebreaker.nim]: We need to find the cause for this though. A refactoring shouldn't have this visible change in allocation behaviour.

File: tests/async/tasync_traceback.nim [hunk 13]
```diff
@@ -110,13 +110,15 @@ Async traceback:
       ## Executes pending callbacks
     asyncmacro\.nim\(\d+?\)\s+?fooNimAsyncContinue
       ## Resumes an async procedure
-    tasync_traceback\.nim\(\d+?\)\s+?fooIter
+    asyncmacro\.nim\(\d+?\)\s+?fooIter
```
Comment (dom96) [inline tests/async/tasync_traceback.nim:113]: This actually reduces the amount of signal we get from the async tracebacks. I cannot recall whether this particular line provided useful information, but if it does then that could be a blocker for these changes. Can you run this test and show us what the tracebacks are before and after your changes?

File: tests/async/tasyncawait_cyclebreaker.nim (hunk 12 above)
Comment (alehander92) [reply tests/async/tasyncawait_cyclebreaker.nim]: yes! I agree

File: lib/pure/asyncmacro.nim (hunk 9 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: it won't as you probably see, because it would lead to an error : await expected Future[T] etc

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: ah, you mean outside of `async`?

File: lib/pure/asyncmacro.nim (hunk 11 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: in order for await to be able to check for this symbol in when IIRC

File: lib/pure/asyncmacro.nim (hunk 9 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: hm,ok, i will try

File: tests/async/tasync_traceback.nim (hunk 13 above)
Comment (alehander92) [reply tests/async/tasync_traceback.nim:113]: IIRC the change is similar to what is seen in the diff, basically we need to add the linedir of `tasync_traceback` to the result of template somehow maybe?

File: lib/pure/asyncmacro.nim (hunk 10 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: maybe because of the `FutureBase` type? but this should be cast-able

File: tests/async/tasyncawait_cyclebreaker.nim (hunk 12 above)
Comment (alehander92) [reply tests/async/tasyncawait_cyclebreaker.nim]: yes, sorry, please read the irc logs in https://irclogs.nim-lang.org/16-04-2020.html#13:51:11 : Araq said to investigate it after all

File: lib/pure/asyncmacro.nim (hunk 8 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: i really think this is also elegant: after all you change how this works quickly with high level code flow without any ast surgery. inMultisync should be somehow optimizeable, its needed only on compile time

Comment (dom96) [reply lib/pure/asyncmacro.nim]: A loop which modifies the AST may be more code, but it's also far more deterministic and easier to debug should something go wrong. Hygiene is important and generating variables so that you can then check for their existence is flaky. The current multisync macro does what it needs to do, and then gets out of your way.

File: tests/async/tasync_traceback.nim (hunk 13 above)
Comment (dom96) [reply tests/async/tasync_traceback.nim:113]: I want to see what line the traceback points to now.

File: lib/pure/asyncmacro.nim (hunk 11 above)
Comment (dom96) [reply lib/pure/asyncmacro.nim]: Ahh. Yeah, I'm not the biggest fan of this. If someone accidentally defines a `retFuture` in their proc then it'll be considered async (that behaviour will be very confusing).

I wonder if we can come up with a better way to detect async procs, can you check if `compiles(yield f)` works? :)

File: lib/pure/asyncmacro.nim (hunk 8 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: but .. wouldn't this break templates again? after all you wouldn't be able to generate await in multisync, i think

File: lib/pure/asyncmacro.nim (hunk 11 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: interesting idea, but this would work in another iterator as well !  which we still don't want. 

maybe i can generate a symbol which can't be written by users similar to `:env`

File: lib/pure/asyncmacro.nim (hunk 9 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: changed

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: changed: i'll wait for the tests

File: lib/pure/asyncmacro.nim (hunk 11 above)
Comment (dom96) [reply lib/pure/asyncmacro.nim]: > but this would work in another iterator as well ! which we still don't want.

IMO it's not the worst thing if it does work, it'll only work if that iterator's return type is a future and if a future is awaited, so likelihood is that it will make sense anyway.

File: lib/pure/asyncmacro.nim (hunk 8 above)
Comment (dom96) [reply lib/pure/asyncmacro.nim]: oh, you mean if a template using `await` is used in a multisync proc? I think you're right. But in that case, can we resolve that in a separate PR? I just like to keep this clean.

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: ok, but we really shouldn't get into a situation where just one of the cases is fixed.. they are symmetric

File: lib/pure/asyncmacro.nim (hunk 11 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: it is far fetched, but it's possible to write a sync iterator returning futures: it makes sense in some cases, so i don't think this would be correct

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: also, i am a bit scared of the `compiles(yield f)` performance thing: for every await, wouldn't that slow down things a bit

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: we can do something else : we can just leave out `template await` and and generate just one in of `async`: Future[T] working and one in multisync sync: which works for every type: all other cases just fail like normal overloads!

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: this way we can really just use the default nim mechanisms for most of the errors as well (await just isn't defined somewhere, or it just doesnt accepts int somewhere)

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: and if we really want to we can still leave a global overload which errors with a more detailed message as a fallback

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: this works!

Comment (alehander92) [reply lib/pure/asyncmacro.nim]: i'll push a new commit when i test it a bit

File: lib/pure/asyncmacro.nim (hunk 10 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: you're right, i need it but not with `when not declared` it seems it doesnt leak now

File: tests/async/tasyncawait_cyclebreaker.nim (hunk 12 above)
Comment (alehander92) [reply tests/async/tasyncawait_cyclebreaker.nim]: another change seems to lead to normal behavior

Comment (alehander92) [reply tests/async/tasyncawait_cyclebreaker.nim]: this should be 1000 now again

File: tests/async/tasync_traceback.nim (hunk 13 above)
Comment (alehander92) [reply tests/async/tasync_traceback.nim:113]: `asyncmacro.nim(261)  `

Comment (alehander92) [reply tests/async/tasync_traceback.nim:113]: which seems to be `(cast[type(f)](internalTmpFuture)).read()`

File: lib/pure/asyncmacro.nim (hunk 8 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim]: ok, no we dont need those variables anyway, so i think this can be still one PR

**Comment by alehander92** [comment]
ok, so now we generate `await` templates inline to not need those retFuture / inMultisync idents, however this require await to be `mixin` in templates using await (or to annotate them with `{.dirty.}`)

this *can* be hinted in an error message tho

**Comment by alehander92** [comment]
sorry! it seems we can just skip the global overload fallback with a custom error and then template without annotation `dirty` can work. not perfect, but i think it's good enough!

File: tests/async/tasync_traceback.nim (hunk 13 above)
Comment (dom96) [reply tests/async/tasync_traceback.nim:113]: That should be okay, I'm assuming that `tasync_traceback` in the old traceback doesn't point to anything interesting though.

File: lib/pure/asyncmacro.nim [hunk 14]
```diff
@@ -332,8 +253,24 @@ proc asyncSingleProc(prc: NimNode): NimNode {.compileTime.} =
     if returnType.kind == nnkEmpty:
       # Add Future[void]
       result.params[0] = parseExpr("owned(Future[void])")
+
+  # based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
```
Comment (dom96) [inline lib/pure/asyncmacro.nim:257]: I would remove this comment, at this point this really doesn't have many parallels. I doubt it'll be useful for those exploring this code.

File: lib/pure/asyncmacro.nim [hunk 15]
```diff
@@ -447,8 +344,9 @@ macro multisync*(prc: untyped): untyped =
...
+  # echo result.repr
 
-proc await*[T](x: T) =
-  ## The 'await' keyword is also defined here for technical
-  ## reasons. (Generic symbol lookup prepass.)
-  {.error: "Await only available within .async".}
+# overload for await as a fallback handler, based on the yglukhov's patch to chronos: https://github.com/status-im/nim-chronos/pull/47
+# template await*(f: typed): untyped =
+  # static:
+    # error "await only available within {.async.}"
```
Comment (dom96) [inline lib/pure/asyncmacro.nim:352]: You still need this, no?

File: lib/pure/asyncmacro.nim [hunk 16]
```diff
@@ -426,8 +319,12 @@ proc splitProc(prc: NimNode): (NimNode, NimNode) =
...
     # Sync proc (0) -> FormalParams (3) -> IdentDefs, the parameter (i) ->
     # parameter type (1).
-    result[0][3][i][1] = splitParamType(result[0][3][i][1], async = false)
-  result[0][6] = stripAwait(result[0][6])
+    result[0][3][i][1] = splitParamType(result[0][3][i][1], async=false)
+  var multisyncAwait = quote:
+    template await(value: typed): untyped =
+      value
+
+  result[0][^1] = nnkStmtList.newTree(multisyncAwait, result[0][^1])
```
Comment (dom96) [inline lib/pure/asyncmacro.nim:327]: I love it :D

File: lib/pure/asyncmacro.nim (hunk 15 above)
Comment (dom96) [reply lib/pure/asyncmacro.nim:352]: oh, you removed this because you didn't want `dirty`. Is `dirty` really that bad? :)

File: lib/pure/asyncmacro.nim (hunk 16 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim:327]: credit to @zah's idea https://github.com/nim-lang/Nim/issues/11911#issuecomment-519410464 i think i finally remembered this is also possible

File: lib/pure/asyncmacro.nim (hunk 14 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim:257]: well, I still got the "define await as a template" overall approach from there, but I am not sure what's the best way to document that stuff

File: lib/pure/asyncmacro.nim (hunk 15 above)
Comment (alehander92) [reply lib/pure/asyncmacro.nim:352]: no, but it can be a bit surprising: we need to somehow hint here in the error "if you use a template, you might want to use `{.dirty.}`" because otherwise many users would be just confused by why just a template/macro can't work 

on the other hand just a type mismatch/undefined error is very natural, `await` is not found and that's it

**Comment by alehander92** [comment]
> I love it, but please check with @Araq that this won't slow down compilation. Generating so many copies of the same template seems like it could go badly :)

Ok, I was wondering about that as well, maybe the compiler should be able to somehow cache such cases (applicable for other template/macro applications)

**Comment by Araq** [comment]
I don't see the copies here, what do you mean?

**Comment by alehander92** [comment]
Araq, it's about inserting two `await` templates into each `async` function (and one into each sync generated by multisync)

**Comment by Araq** [comment]
I cannot imagine these 2 little templates are noticable, merging.
